# Notebook: AISC2009_CaseStudy1_Group1.ipynb

2. Perform lemmatization using NLTK lemmatizer. Count the number of unique words/tokens before and after lemmatization


In [29]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer
from nltk.corpus import stopwords


In [31]:
nltk.download('wordnet') # dictionary
nltk.download('stopwords') # stop words

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\sspc1\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\sspc1\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [37]:
data = pd.read_csv("E:/Subhash_pavan/CANADA/loyalist/classes/SEMESTER3/NLP/projects/Group1/Case_Study1/src/notebook/data/Data.csv")

In [11]:
data.head()

,Data
0,Watch or listen live weekdays at 8:30am MT at ...
1,Watch or listen live weekdays at 8:30am MT at ...
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
4,"Journalist, publisher of Rebel News — telling ..."


In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7339 entries, 0 to 7338
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Data    7339 non-null   object
dtypes: object(1)
memory usage: 57.5+ KB


In [16]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth',None)
data.head()

,Data
0,Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
1,Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
4,"Journalist, publisher of Rebel News — telling the other side of the story. Awarded the Queen’s Diamond Jubilee Medal for advancing freedom of expression."


In [42]:
text_data="".join(data['Data'].astype(str)).lower()
text_data

'watch or listen live weekdays at 8:30am mt at ryanjespersen.com. subscribe via youtube or your favourite podcast app. #realtalkrjwatch or listen live weekdays at 8:30am mt at ryanjespersen.com. subscribe via youtube or your favourite podcast app. #realtalkrjchubby and hot, always stir the pot!chubby and hot, always stir the pot!journalist, publisher of rebel news — telling the other side of the story. awarded the queen’s diamond jubilee medal for advancing freedom of expression.edmonton city councillor for #wardmétis. #yeg urban planner, sustainability & sociologist. founder. cities, climate action, health, inclusion. she/her #yegccalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notley and staff. she/her #ablegalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notley and staff. she/her #ablegalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notl

In [41]:
text_data=re.sub(r'\b\w+\.com\b', '', text_data)
text_data

'watch or listen live weekdays at 8:30am mt at . subscribe via youtube or your favourite podcast app. #realtalkrjwatch or listen live weekdays at 8:30am mt at . subscribe via youtube or your favourite podcast app. #realtalkrjchubby and hot, always stir the pot!chubby and hot, always stir the pot!journalist, publisher of rebel news — telling the other side of the story. awarded the queen’s diamond jubilee medal for advancing freedom of expression.edmonton city councillor for #wardmétis. #yeg urban planner, sustainability & sociologist. founder. cities, climate action, health, inclusion. she/her #yegccalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notley and staff. she/her #ablegalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notley and staff. she/her #ablegalbertan. mom to children & dogs. wife. friend. leader of the alberta ndp. account run by rachel notley and staff. she/her #ablegalbert

In [43]:
# Removing unwanted characters

text_data = re.sub(r'http\S+|www\S+|https\S+', '', text_data) # remove URLs
text_data=re.sub(r'\b\w+\.com\b', '', text_data)
text_data = re.sub(r'\b\w+\.co\b', '', text_data)
text_data = re.sub(r'\b\w+\.org\b', '', text_data)  # remove domain names
text_data = re.sub(r'\b\w+\.net\b', '', text_data)
text_data = re.sub(r'\S+@\S+', '', text_data) # remove email addresses
text_data = re.sub(r'[^a-zA-Z\s]', '', text_data) # remove special characters
text_data = re.sub(r'<.*?>', '', text_data)  # remove HTML tags
text_data = re.sub(r'\d+', '', text_data)  # remove numbers
text_data = re.sub(r'\s+', ' ', text_data)  # remove extra spaces

text_data

'watch or listen live weekdays at am mt at subscribe via youtube or your favourite podcast app realtalkrjwatch or listen live weekdays at am mt at subscribe via youtube or your favourite podcast app realtalkrjchubby and hot always stir the potchubby and hot always stir the potjournalist publisher of rebel news telling the other side of the story awarded the queens diamond jubilee medal for advancing freedom of expressionedmonton city councillor for wardmtis yeg urban planner sustainability sociologist founder cities climate action health inclusion sheher yegccalbertan mom to children dogs wife friend leader of the alberta ndp account run by rachel notley and staff sheher ablegalbertan mom to children dogs wife friend leader of the alberta ndp account run by rachel notley and staff sheher ablegalbertan mom to children dogs wife friend leader of the alberta ndp account run by rachel notley and staff sheher ablegalbertan mom to children dogs wife friend leader of the alberta ndp account r

In [44]:
# Tokenization
tokens = re.findall(r'\b[a-z]+\b', text_data)
tokens

['watch',
 'or',
 'listen',
 'live',
 'weekdays',
 'at',
 'am',
 'mt',
 'at',
 'subscribe',
 'via',
 'youtube',
 'or',
 'your',
 'favourite',
 'podcast',
 'app',
 'realtalkrjwatch',
 'or',
 'listen',
 'live',
 'weekdays',
 'at',
 'am',
 'mt',
 'at',
 'subscribe',
 'via',
 'youtube',
 'or',
 'your',
 'favourite',
 'podcast',
 'app',
 'realtalkrjchubby',
 'and',
 'hot',
 'always',
 'stir',
 'the',
 'potchubby',
 'and',
 'hot',
 'always',
 'stir',
 'the',
 'potjournalist',
 'publisher',
 'of',
 'rebel',
 'news',
 'telling',
 'the',
 'other',
 'side',
 'of',
 'the',
 'story',
 'awarded',
 'the',
 'queens',
 'diamond',
 'jubilee',
 'medal',
 'for',
 'advancing',
 'freedom',
 'of',
 'expressionedmonton',
 'city',
 'councillor',
 'for',
 'wardmtis',
 'yeg',
 'urban',
 'planner',
 'sustainability',
 'sociologist',
 'founder',
 'cities',
 'climate',
 'action',
 'health',
 'inclusion',
 'sheher',
 'yegccalbertan',
 'mom',
 'to',
 'children',
 'dogs',
 'wife',
 'friend',
 'leader',
 'of',
 'the',

In [45]:
# Stopword Removal
stop_words = set(stopwords.words('english'))

final_tokens = [word for word in tokens if word not in stop_words]
final_tokens

['watch',
 'listen',
 'live',
 'weekdays',
 'mt',
 'subscribe',
 'via',
 'youtube',
 'favourite',
 'podcast',
 'app',
 'realtalkrjwatch',
 'listen',
 'live',
 'weekdays',
 'mt',
 'subscribe',
 'via',
 'youtube',
 'favourite',
 'podcast',
 'app',
 'realtalkrjchubby',
 'hot',
 'always',
 'stir',
 'potchubby',
 'hot',
 'always',
 'stir',
 'potjournalist',
 'publisher',
 'rebel',
 'news',
 'telling',
 'side',
 'story',
 'awarded',
 'queens',
 'diamond',
 'jubilee',
 'medal',
 'advancing',
 'freedom',
 'expressionedmonton',
 'city',
 'councillor',
 'wardmtis',
 'yeg',
 'urban',
 'planner',
 'sustainability',
 'sociologist',
 'founder',
 'cities',
 'climate',
 'action',
 'health',
 'inclusion',
 'sheher',
 'yegccalbertan',
 'mom',
 'children',
 'dogs',
 'wife',
 'friend',
 'leader',
 'alberta',
 'ndp',
 'account',
 'run',
 'rachel',
 'notley',
 'staff',
 'sheher',
 'ablegalbertan',
 'mom',
 'children',
 'dogs',
 'wife',
 'friend',
 'leader',
 'alberta',
 'ndp',
 'account',
 'run',
 'rachel',

In [46]:
#unique words before Lemmatization
unique_words_before_lemmatization = set(final_tokens)
len(unique_words_before_lemmatization)

9782

In [47]:
# Lemmatization

Lemmatizer=WordNetLemmatizer()
lemmas=[Lemmatizer.lemmatize(word) for word in final_tokens]
lemmas

['watch',
 'listen',
 'live',
 'weekday',
 'mt',
 'subscribe',
 'via',
 'youtube',
 'favourite',
 'podcast',
 'app',
 'realtalkrjwatch',
 'listen',
 'live',
 'weekday',
 'mt',
 'subscribe',
 'via',
 'youtube',
 'favourite',
 'podcast',
 'app',
 'realtalkrjchubby',
 'hot',
 'always',
 'stir',
 'potchubby',
 'hot',
 'always',
 'stir',
 'potjournalist',
 'publisher',
 'rebel',
 'news',
 'telling',
 'side',
 'story',
 'awarded',
 'queen',
 'diamond',
 'jubilee',
 'medal',
 'advancing',
 'freedom',
 'expressionedmonton',
 'city',
 'councillor',
 'wardmtis',
 'yeg',
 'urban',
 'planner',
 'sustainability',
 'sociologist',
 'founder',
 'city',
 'climate',
 'action',
 'health',
 'inclusion',
 'sheher',
 'yegccalbertan',
 'mom',
 'child',
 'dog',
 'wife',
 'friend',
 'leader',
 'alberta',
 'ndp',
 'account',
 'run',
 'rachel',
 'notley',
 'staff',
 'sheher',
 'ablegalbertan',
 'mom',
 'child',
 'dog',
 'wife',
 'friend',
 'leader',
 'alberta',
 'ndp',
 'account',
 'run',
 'rachel',
 'notley',
 

In [36]:
#unique words after Lemmatization
unique_words_after_lemmatization = set(lemmas)
len(unique_words_after_lemmatization)

9288

In [48]:

print("Number of unique words before lemmatization:", len(unique_words_before_lemmatization))
print("Number of unique words after lemmatization:", len(unique_words_after_lemmatization))




Number of unique words before lemmatization: 9782
Number of unique words after lemmatization: 9288


# Notebook: AISC2009_CaseStudy1_Group1_PorterStemming.ipynb

## Imports

In [82]:
import pandas as pd
import re
from nltk.tokenize import TweetTokenizer
from nltk.stem import PorterStemmer

## Reading Data

In [83]:
df = pd.read_csv("Data.csv")


In [84]:
df.head

<bound method NDFrame.head of                                                    Data
0     Watch or listen live weekdays at 8:30am MT at ...
1     Watch or listen live weekdays at 8:30am MT at ...
2                  Chubby And Hot, Always Stir The Pot!
3                  Chubby And Hot, Always Stir The Pot!
4     Journalist, publisher of Rebel News — telling ...
...                                                 ...
7334  Snow Valley Racing delivers competitive ski ra...
7335  New twin mom\n\nMorning News Anchor @citynewsc...
7336  Residents advocating for the💚of the Park.\n\nP...
7337  Driven by family, diversity, equity, inclusion...
7338      Photographer, Oilers Fan #photography  #yeg .

[7339 rows x 1 columns]>

## Concatenate all the text entries

In [85]:
text_data = " ".join(df['Data'].dropna().astype(str))

## Tokenize data

In [86]:
tokens = re.findall(r'\b[a-zA-Z]+\b', text_data.lower())

In [87]:
unique_before = set(tokens)


## Perform Stemming

In [88]:
stemmer = PorterStemmer()
stemmed_tokens = [stemmer.stem(token) for token in tokens]

In [89]:
unique_after = set(stemmed_tokens)

In [90]:
print("Unique tokens before stemming:", len(unique_before))
print("Unique tokens after stemming:", len(unique_after))

Unique tokens before stemming: 7445
Unique tokens after stemming: 6088


In [91]:
print("First 20 stemmed tokens:")
print(stemmed_tokens[:20])

First 20 stemmed tokens:
['watch', 'or', 'listen', 'live', 'weekday', 'at', 'mt', 'at', 'ryanjespersen', 'com', 'subscrib', 'via', 'youtub', 'or', 'your', 'favourit', 'podcast', 'app', 'realtalkrj', 'watch']


### While doing this method, the stemmer is skipping numbers. So, to treat them. 

In [79]:
tokens = re.findall(r"\b\w[\w'.:/@-]*\b", text_data)
unique_before = set(tokens)

In [80]:
stemmed_tokens = [
    stemmer.stem(token) if token.isalpha() else token
    for token in tokens
]

unique_after = set(stemmed_tokens)

In [81]:
print("Unique tokens before stemming:", len(unique_before))
print("Unique tokens after stemming:", len(unique_after))
print("First 20 stemmed tokens:")
print(stemmed_tokens[:20])

Unique tokens before stemming: 10056
Unique tokens after stemming: 7033
First 20 stemmed tokens:
['watch', 'or', 'listen', 'live', 'weekday', 'at', '8:30am', 'mt', 'at', 'ryanjespersen.com', 'subscrib', 'via', 'youtub', 'or', 'your', 'favourit', 'podcast', 'app', 'realtalkrj', 'watch']


# Notebook: CaseStudy1_Q5.ipynb

In [ ]:
!pip install spacy

In [2]:
!python -m spacy download en_core_web_sm


     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     - -------------------------------------- 0.5/12.8 MB 2.8 MB/s eta 0:00:05
     ----- ---------------------------------- 1.8/12.8 MB 5.6 MB/s eta 0:00:02
     ----------- ---------------------------- 3.7/12.8 MB 5.6 MB/s eta 0:00:02
     ----------------- ---------------------- 5.5/12.8 MB 6.6 MB/s eta 0:00:02
     ---------------------- ----------------- 7.3/12.8 MB 7.1 MB/s eta 0:00:01
     ----------------------------- ---------- 9.4/12.8 MB 7.4 MB/s eta 0:00:01
     ---------------------------------- ----- 11.0/12.8 MB 7.5 MB/s eta 0:00:01
     ---------------------------------------  12.6/12.8 MB 7.5 MB/s eta 0:00:01
     ---------------------------------------- 12.8/12.8 MB 7.4 MB/s eta 0:00:00
[+] Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [3]:
import re
import pandas as pd
# Load the uploaded CSV file
file_path = 'Data.csv'
df = pd.read_csv(file_path)

# Display the first few rows and the column names to understand the structure
df.head(), df.columns

(                                                Data
 0  Watch or listen live weekdays at 8:30am MT at ...
 1  Watch or listen live weekdays at 8:30am MT at ...
 2               Chubby And Hot, Always Stir The Pot!
 3               Chubby And Hot, Always Stir The Pot!
 4  Journalist, publisher of Rebel News — telling ...,
 Index(['Data'], dtype='object'))

In [4]:

# Show full column content when printing
pd.set_option('display.max_colwidth', None)  # Removes truncation of long strings

# Define regex pattern to match decimal numbers like "5."
pattern_5_dot = r'\b\d+\.\b'

# Find matching rows
match_5_dot = df[df['Data'].str.contains(pattern_5_dot, regex=True)]

# Display matching records
print("Matching records:\n")
print(match_5_dot)

# Display row indices
print("\nLine numbers (0-based index) with '5.':")
print(match_5_dot.index.tolist())


Matching records:

                                                                                                                                                                   Data
473                                       City of Edmonton News & Updates. Active weekdays 8am-4:30pm. Report concerns: 311.edmonton.ca. Edmonton Transit @takeETSalert
487                                       City of Edmonton News & Updates. Active weekdays 8am-4:30pm. Report concerns: 311.edmonton.ca. Edmonton Transit @takeETSalert
2388         Don Hladiuk is Calgary's eye on astronomy & space science news. On the CBC EyeOpener 1010AM or 99.1FM on the 1st or 2nd Monday of the month at 7:36 AM MT.
2718         Don Hladiuk is Calgary's eye on astronomy & space science news. On the CBC EyeOpener 1010AM or 99.1FM on the 1st or 2nd Monday of the month at 7:36 AM MT.
3208                                     AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest 

In [5]:
duplicate_count = df.duplicated().sum()
print("Number of completely duplicate rows:", duplicate_count)


Number of completely duplicate rows: 5353


In [6]:
text_duplicates = df['Data'].duplicated().sum()
print("Number of duplicate 'Data' entries:", text_duplicates)


Number of duplicate 'Data' entries: 5353


In [7]:
duplicated_rows = df[df['Data'].duplicated(keep=False)]
duplicated_rows.head()


,Data
0,Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
1,Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
6,Albertan. Mom to children & dogs. Wife. Friend. Leader of the Alberta NDP. Account run by Rachel Notley and staff. She/Her #ableg


In [8]:
df_no_duplicates = df.drop_duplicates(subset='Data').copy()


These are dates, not decimal numbers.

You may want to exclude dates from decimal detection if accuracy is important for numerical feature extraction.

Standard decimals: 12.5, .75

Numbers ending with a dot: 5.

Comma-formatted decimals: 5,000.25

Scientific notation: 1.2e3, 2E5


Pre-tag or remove timestamps/dates with regex

In [20]:


# Regex patterns (as you have)
time_pattern = r'\b\d{1,2}:\d{2}(?:[ap]m)?\b'
date_pattern = r'\b\d{1,2}[./-]\d{1,2}[./-]?\d{0,4}\b'
year_range_pattern = r'\b\d{4}[-–]\d{2,4}\b'
apostrophe_year = r"\b'\d{2}\b"

combined_pattern = f'({time_pattern}|{date_pattern}|{year_range_pattern}|{apostrophe_year})'

# Use .loc for assignment to avoid SettingWithCopyWarning
df_no_duplicates.loc[:, 'Cleaned_Data'] = df_no_duplicates['Data'].str.replace(combined_pattern, '', regex=True)

decimal_pattern = r'\b\d*\.\d+\b|\b\d+\.\b|\b\d{1,3}(?:,\d{3})*(?:\.\d+)?\b|\b\d+\.?\d*[eE][+-]?\d+\b'
df_no_duplicates.loc[:, 'Decimals'] = df_no_duplicates['Cleaned_Data'].str.findall(decimal_pattern)

print(df_no_duplicates[['Data', 'Decimals']].head())



                                                                                                                                                            Data  \
0                              Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ   
2                                                                                                                           Chubby And Hot, Always Stir The Pot!   
4      Journalist, publisher of Rebel News — telling the other side of the story. Awarded the Queen’s Diamond Jubilee Medal for advancing freedom of expression.   
5  Edmonton City Councillor for #WardMétis. #YEG Urban Planner, Sustainability & Sociologist. Founder. Cities, climate action, health, inclusion. She/her #yegcc   
6                              Albertan. Mom to children & dogs. Wife. Friend. Leader of the Alberta NDP. Account run by Rachel Notley and staff. She/Her #ableg   

  Decimals  
0 

 Use Named Entity Recognition (NER)

In [21]:
import spacy

# Load English NER model
nlp = spacy.load("en_core_web_sm")

def remove_dates_and_times(text):
    doc = nlp(text)
    cleaned = ''.join([
        token.text_with_ws for token in doc
        if token.ent_type_ not in ["DATE", "TIME"]
    ])
    return cleaned

# Safely apply to dataset
df_no_duplicates.loc[:, 'NER_Cleaned'] = df_no_duplicates['Data'].apply(remove_dates_and_times)

# Apply decimal regex after NER cleaning (use .loc)
df_no_duplicates.loc[:, 'NER_Decimals'] = df_no_duplicates['NER_Cleaned'].str.findall(decimal_pattern)


In [22]:
# Regex pattern to match comma-formatted decimals like "5,000.25"
pattern = r'\b\d{1,3}(?:,\d{3})+\.\d+\b'

# Filter rows that match the pattern
matches = df_no_duplicates[df_no_duplicates['Data'].str.contains(pattern, regex=True)]

# Print results
print("Total matches:", len(matches))
print(matches[['Data']])

Total matches: 0
Empty DataFrame
Columns: [Data]
Index: []


In [23]:
# Regex pattern to match scientific notation (e.g., 1.2e3, 2E5, 3.0E-2)
pattern = r'\b\d+\.?\d*[eE][+-]?\d+\b'

# Filter rows that match the pattern
matches = df_no_duplicates[df_no_duplicates['Data'].str.contains(pattern, regex=True)]

# Print results
print("Total matches:", len(matches))
print(matches[['Data']])

Total matches: 0
Empty DataFrame
Columns: [Data]
Index: []


In [24]:
# Make sure these patterns are defined before combining!
time_pattern = r'\b\d{1,2}:\d{2}(?:[ap]m)?\b'
date_pattern = r'\b\d{1,2}[./-]\d{1,2}(?:[./-]\d{2,4})?\b'
year_range_pattern = r'\b\d{4}[-–]\d{2,4}\b'
apostrophe_year = r"\b'\d{2}\b"
phone_pattern = r'\+?\d{1,2}\.\d{3}\.\d{3}\.\d{4}'

# Combine all patterns into one
combined_pattern = f"({time_pattern}|{date_pattern}|{year_range_pattern}|{apostrophe_year}|{phone_pattern})"

# Clean those patterns out of the text
df_no_duplicates.loc[:, 'Cleaned_Data'] = df_no_duplicates['Data'].str.replace(combined_pattern, '', regex=True)

# Final extended decimal regex:
decimal_pattern = r'''
    \b\d*\.\d+\b                    # standard decimal like .75, 12.3
    | \b\d{1,3}(?:,\d{3})+\.\d+\b   # comma formatted decimal like 5,000.25
    | \b\d+\.?\d*[eE][+-]?\d+\b     # scientific notation like 1.2e3, 2E5
'''

# Check which rows contain a decimal
contains_decimal = df_no_duplicates['Cleaned_Data'].str.contains(decimal_pattern, regex=True, flags=re.VERBOSE)

# Count matching records
decimal_count = contains_decimal.sum()
print(f"Number of records containing a decimal number: {decimal_count}")



Number of records containing a decimal number: 3


In [25]:
# Display the 17 matching records
matching_rows = df_no_duplicates[contains_decimal]
print(matching_rows[['Data']])


                                                                                                                                                                 Data
3208                                   AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC
4568  #openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: 3.2.\n#LULC #landscape #soil #urban #climate
5113                                                                                                  Playing Southern Saskatchewan's Best Country! Country 100.7 FM!


In [26]:
# Regex to find decimal numbers
decimal_pattern = r'\b\d+\.\d+\b'

# Count records containing decimal numbers
decimal_count = df_no_duplicates['Data'].str.contains(decimal_pattern, regex=True).sum()

print("Number of records with a decimal number:", decimal_count)
# Filter rows with decimal numbers
decimal_rows = df_no_duplicates[df_no_duplicates['Data'].str.contains(r'\b\d+\.\d+\b', regex=True)]

# Display them
decimal_rows

Number of records with a decimal number: 7


,Data,Cleaned_Data,Decimals,NER_Cleaned,NER_Decimals
3208,AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC,AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC,[106.1],AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC,[106.1]
3663,hier schreiben die Hühner aus dem Stall (*25./26.4. 2019) . \nKörner und Würmer immer willkommen,hier schreiben die Hühner aus dem Stall (*25./. 2019) . \nKörner und Würmer immer willkommen,[25],hier schreiben die Hühner aus dem Stall (*) . \nKörner und Würmer immer willkommen,[]
3682,"Providing RV repair, parts & accessories, 24/7 mobile service, custom rebuilds, rentals & quality advice from Red Seal Certified RV Technicians. +1.587.989.1271","Providing RV repair, parts & accessories, mobile service, custom rebuilds, rentals & quality advice from Red Seal Certified RV Technicians.",[989.1271],"Providing RV repair, parts & accessories, 24/7 mobile service, custom rebuilds, rentals & quality advice from Red Seal Certified RV Technicians. +1.587.989.1271","[24, 7, 1.587, .989, .1271]"
4568,"#openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: 3.2.\n#LULC #landscape #soil #urban #climate","#openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: .\n#LULC #landscape #soil #urban #climate",[],"#openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF : 3.905; CiteScore : 3.2.\n#LULC #landscape #soil #urban #climate","[3.905, 3.2]"
4626,"Hoping hardcore twitter 2.0 lets me express honest opinions about Danielle Smith's supporters.\n\n""Let them see what is on the end of that long newspaper spoon.""","Hoping hardcore twitter lets me express honest opinions about Danielle Smith's supporters.\n\n""Let them see what is on the end of that long newspaper spoon.""",[],"Hoping hardcore twitter 2.0 lets me express honest opinions about Danielle Smith's supporters.\n\n""Let them see what is on the end of that long newspaper spoon.""",[2.0]
5113,Playing Southern Saskatchewan's Best Country! Country 100.7 FM!,Playing Southern Saskatchewan's Best Country! Country 100.7 FM!,[100.7],Playing Southern Saskatchewan's Best Country! Country 100.7 FM!,[100.7]
5244,96.5 CKFM - All Hit Country!\n📻 96.5 FM \n🌐 ckfm.ca \nYour source for everything happening in Mountain View County!,CKFM - All Hit Country!\n📻 FM \n🌐 ckfm.ca \nYour source for everything happening in Mountain View County!,[],96.5 CKFM - All Hit Country!\n📻 96.5 FM \n🌐 ckfm.ca \nYour source for everything happening in Mountain View County!,"[96.5, 96.5]"


In [29]:
# Step 1: Define cleanup patterns
time_pattern = r'\b\d{1,2}:\d{2}(?:[ap]m)?\b'
date_pattern = r'\b\d{1,2}[./-]\d{1,2}(?:[./-]\d{2,4})?\b'
year_range_pattern = r'\b\d{4}[-–]\d{2,4}\b'
apostrophe_year = r"\b'\d{2}\b"
phone_pattern = r'\+?\d{1,2}\.\d{3}\.\d{3}\.\d{4}'

# Combine all patterns
combined_pattern = f"({time_pattern}|{date_pattern}|{year_range_pattern}|{apostrophe_year}|{phone_pattern})"

# Step 2: Clean the Data column using .loc to avoid SettingWithCopyWarning
df_no_duplicates.loc[:, 'Cleaned_Data'] = df_no_duplicates['Data'].str.replace(combined_pattern, '', regex=True)

# Step 3: Define decimal detection pattern (excluding cleaned cases)
decimal_pattern = r'''
    \b\d*\.\d+\b                    # standard decimal like .75, 12.3
    | \b\d{1,3}(?:,\d{3})+\.\d+\b   # comma formatted decimal like 5,000.25
    | \b\d+\.?\d*[eE][+-]?\d+\b     # scientific notation like 1.2e3, 2E5
'''

# Step 4: Detect decimals in cleaned data
mask = df_no_duplicates['Cleaned_Data'].str.contains(decimal_pattern, regex=True, flags=re.VERBOSE)
decimal_count = mask.sum()
matching_rows = df_no_duplicates[mask]

# Step 5: Output the results
print("✅ Number of records with decimal numbers after cleaning:", decimal_count)
print("\n✅ Matching records:")
print(matching_rows[['Data']])




✅ Number of records with decimal numbers after cleaning: 3

✅ Matching records:
                                                                                                                                                                 Data
3208                                   AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC
4568  #openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: 3.2.\n#LULC #landscape #soil #urban #climate
5113                                                                                                  Playing Southern Saskatchewan's Best Country! Country 100.7 FM!


In [31]:
# Step 1: Define cleanup patterns (excluding date to preserve decimals like 96.5)
time_pattern = r'\b\d{1,2}:\d{2}(?:[ap]m)?\b'
year_range_pattern = r'\b\d{4}[-–]\d{2,4}\b'
apostrophe_year = r"\b'\d{2}\b"
phone_pattern = r'\+?\d{1,2}\.\d{3}\.\d{3}\.\d{4}(?!\d)'

# Step 2: Combine only safe cleaning patterns
combined_pattern = f"({time_pattern}|{year_range_pattern}|{apostrophe_year}|{phone_pattern})"

# Step 3: Clean the Data column - FIXED with .loc to prevent warnings
df_no_duplicates.loc[:, 'Cleaned_Data'] = df_no_duplicates['Data'].str.replace(combined_pattern, '', regex=True)

# Step 4: Define decimal detection pattern (standard, comma, scientific)
decimal_pattern = r'''
    \b\d*\.\d+\b                    # .75, 12.3
    | \b\d{1,3}(?:,\d{3})+\.\d+\b   # 5,000.25
    | \b\d+\.?\d*[eE][+-]?\d+\b     # 1.2e3, 2E5
'''

# Step 5: Detect decimals in cleaned data (no warning needed here, only for assignments)
mask = df_no_duplicates['Cleaned_Data'].str.contains(decimal_pattern, regex=True, flags=re.VERBOSE)
decimal_count = mask.sum()
matching_rows = df_no_duplicates[mask]

# Step 6: Output the results
print("✅ Number of records with decimal numbers after cleaning:", decimal_count)
print("\n✅ Matching records:")
print(matching_rows[['Data']])



✅ Number of records with decimal numbers after cleaning: 6

✅ Matching records:
                                                                                                                                                                   Data
3208                                     AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC
3663                                                                  hier schreiben die Hühner aus dem Stall  (*25./26.4. 2019) . \nKörner und Würmer immer willkommen
4568    #openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: 3.2.\n#LULC #landscape #soil #urban #climate
4626  Hoping hardcore twitter 2.0 lets me express honest opinions about Danielle Smith's supporters.\n\n"Let them see what is on the end of that long newspaper spoon."
5113                                                                            

The date pattern is included, when we try to reomve, it removes the real values like 96.5 hence needed trying to create a custom date pattern.

In [32]:
# Step 1: Define cleanup patterns (excluding date to preserve decimals like 96.5)
time_pattern = r'\b\d{1,2}:\d{2}(?:[ap]m)?\b'
year_range_pattern = r'\b\d{4}[-–]\d{2,4}\b'
apostrophe_year = r"\b'\d{2}\b"
phone_pattern = r'\+?\d{1,2}\.\d{3}\.\d{3}\.\d{4}(?!\d)'

# Step 2: Combine only safe cleaning patterns
combined_pattern = f"({time_pattern}|{year_range_pattern}|{apostrophe_year}|{phone_pattern})"

# Step 3: Clean the Data column - FIXED with .loc to prevent warnings
df_no_duplicates.loc[:, 'Cleaned_Data'] = df_no_duplicates['Data'].str.replace(combined_pattern, '', regex=True)

# Step 4: Define decimal detection pattern (standard, comma, scientific)
decimal_pattern = r'''
    \b\d*\.\d+\b                    # .75, 12.3
    | \b\d{1,3}(?:,\d{3})+\.\d+\b   # 5,000.25
    | \b\d+\.?\d*[eE][+-]?\d+\b     # 1.2e3, 2E5
'''

# Step 5: Detect decimals in cleaned data (no warning needed here, only for assignments)
mask = df_no_duplicates['Cleaned_Data'].str.contains(decimal_pattern, regex=True, flags=re.VERBOSE)
decimal_count = mask.sum()
matching_rows = df_no_duplicates[mask]

# Step 6: Output the results
print("✅ Number of records with decimal numbers after cleaning:", decimal_count)
print("\n✅ Matching records:")
print(matching_rows[['Data']])


✅ Number of records with decimal numbers after cleaning: 6

✅ Matching records:
                                                                                                                                                                   Data
3208                                     AIR 106.1 FM + DiscoverAirdrie Local News.\nWant to win a truck? Our Ram Everyday Adventure Contest is BACK!\n\n@RadioClaireMC
3663                                                                  hier schreiben die Hühner aus dem Stall  (*25./26.4. 2019) . \nKörner und Würmer immer willkommen
4568    #openaccess journal published by@MDPIOpenAccess, covering aspects of #landscience. IF 2021: 3.905; CiteScore 2021: 3.2.\n#LULC #landscape #soil #urban #climate
4626  Hoping hardcore twitter 2.0 lets me express honest opinions about Danielle Smith's supporters.\n\n"Let them see what is on the end of that long newspaper spoon."
5113                                                                            

<h2>6 records contain a decimal number</h2>

**Reference List (APA 7):**
OpenAI. (2025). ChatGPT (May 17 version) [Large language model]. https://chat.openai.com/  


First Prompt: "How to use regex to find decimal numbers"   
Last Prompt : "records 3663, and 4316 has to be cleaned, without affecting other records as they are date formats, how to handle that"

# Notebook: CaseStudy1_Q7.ipynb

<h1>Q7) How many records have a new line?</h1>

In [15]:
import re
import pandas as pd
# Load the uploaded CSV file
file_path = 'Data.csv'
df = pd.read_csv(file_path)

# Display the first few rows and the column names to understand the structure
df.head(), df.columns

(                                                Data
 0  Watch or listen live weekdays at 8:30am MT at ...
 1  Watch or listen live weekdays at 8:30am MT at ...
 2               Chubby And Hot, Always Stir The Pot!
 3               Chubby And Hot, Always Stir The Pot!
 4  Journalist, publisher of Rebel News — telling ...,
 Index(['Data'], dtype='object'))

In [16]:
duplicate_count = df.duplicated().sum()
print("Number of completely duplicate rows:", duplicate_count)

Number of completely duplicate rows: 5353


In [17]:
text_duplicates = df['Data'].duplicated().sum()
print("Number of duplicate 'Data' entries:", text_duplicates)

Number of duplicate 'Data' entries: 5353


In [18]:
duplicated_rows = df[df['Data'].duplicated(keep=False)]
duplicated_rows.head()


,Data
0,Watch or listen live weekdays at 8:30am MT at ...
1,Watch or listen live weekdays at 8:30am MT at ...
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
6,Albertan. Mom to children & dogs. Wife. Friend...


In [19]:
df_no_duplicates = df.drop_duplicates(subset='Data')


In [20]:
df_no_duplicates = df_no_duplicates.reset_index(drop=True)


In [21]:
newline_pattern = r'\n'


In [26]:
mask = df_no_duplicates['Data'].str.contains(newline_pattern)
newline_count = mask.sum()
matching_rows = df_no_duplicates[mask]


In [27]:
print("✅ Number of records with a newline:", newline_count)
print(matching_rows[['Data']].head())

✅ Number of records with a newline: 299
                                                 Data
7   Think-tank president. Out-of-touch opinion eli...
13  Traveller, nature lover, photographer. HR Prof...
20  📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time ...
24  When I look at the mess that is Alberta under ...
26  Alberta's Most Followed War Room--double the f...


**Factors to consider**
* Multiple newlines \n\n\n	May create tokenization artifacts or unintended paragraph breaks.
* Trailing/leading newlines	Can cause padding issues in prompt formatting.
* HTML <br> used instead	Missed by simple \n regex — should optionally detect.
* Invisibles like \r\n (Windows)	Could lead to false negatives if not normalized.
* Text split mid-sentence	Can affect context chunking or coherence in LLM summarization.


In [30]:
# Define a pattern to match all newline variants
newline_pattern = r'(?:\r\n|\r|\n|<br\s*/?>)'

# Count number of newlines in each record
df_no_duplicates['newline_count'] = df_no_duplicates['Data'].str.count(newline_pattern, flags=re.IGNORECASE)

# Count total records with at least one newline
has_newline_mask = df['newline_count'] >= 1
total_with_newlines = has_newline_mask.sum()

# Count total records with 2 or more newlines
multi_newline_mask = df_no_duplicates['newline_count'] >= 2
multi_newline_count = multi_newline_mask.sum()

# Print results
print(f"✅ Total records with at least one newline: {total_with_newlines}")
print(f"✅ Records with 2 or more newlines: {multi_newline_count}")

# Optional: preview those records
print("\n🔍 Sample of records with 2+ newlines:")
print(df_no_duplicates[multi_newline_mask][['Data', 'newline_count']].head())


✅ Total records with at least one newline: 299
✅ Records with 2 or more newlines: 184

🔍 Sample of records with 2+ newlines:
                                                 Data  newline_count
20  📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time ...              4
24  When I look at the mess that is Alberta under ...              2
42  Ban semi auto rifles w/detachable magazines an...              2
61                                  Teacher\n\nHe/Him              2
73  Greek #Archaeology. State formation, #Thessaly...              2


<h2> 299 records contain a decimal number </h2>

**Reference List (APA 7):**
OpenAI. (2025). ChatGPT (May 17 version) [Large language model]. https://chat.openai.com/  

First Prompt: "How to use regex to find newlines"

Last Prompt : "count how many records have 2 or more newlines? That would help detect paragraphs or long posts, which often require chunking for LLM input"

# Notebook: CaseStudy1_Serene_Q6_Q8.ipynb

## Import Libraries

In [113]:
import pandas as pd
import re
import ipaddress

## Load the dataset

In [115]:
language_data= pd.read_csv("Data.csv")

## Display the dataset

In [117]:
language_data.head(10)

,Data
0,Watch or listen live weekdays at 8:30am MT at ...
1,Watch or listen live weekdays at 8:30am MT at ...
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
4,"Journalist, publisher of Rebel News — telling ..."
5,Edmonton City Councillor for #WardMétis. #YEG ...
6,Albertan. Mom to children & dogs. Wife. Friend...
7,Albertan. Mom to children & dogs. Wife. Friend...
8,Albertan. Mom to children & dogs. Wife. Friend...
9,Albertan. Mom to children & dogs. Wife. Friend...


In [118]:
language_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7339 entries, 0 to 7338
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Data    7339 non-null   object
dtypes: object(1)
memory usage: 57.5+ KB


## Data Preprocessing

Check for null values

In [121]:
print(language_data.isnull().sum())

Data    0
dtype: int64


Check for duplicate values

In [123]:
duplicate_count = language_data.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 5353


In [124]:
language_data = language_data.drop_duplicates()

In [125]:
print("Current shape of dataset:", language_data.shape)

Current shape of dataset: (1986, 1)


In [126]:
duplicate_count = language_data.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 0


## Q.No.  A.6

#### What is the total number of ip addresses across all the records?

In [129]:
text_data = ' '.join(language_data['Data'].dropna().astype(str))

# Regex patterns for different IP formats
ipv4_pattern = r'\b(?:(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\.){3}(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\b'
cidr_pattern = r'\b(?:(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\.){3}(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)/[0-9]{1,2}\b'
ipv6_full = r'\b(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\b'
ipv6_compressed = r'\b(?:[A-Fa-f0-9]{1,4}:){1,7}:|::(?:[A-Fa-f0-9]{1,4}:){1,7}[A-Fa-f0-9]{1,4}\b'
ipv6_mapped_ipv4 = r'\b::ffff:(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\.(?:25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)\b'

# Additional non-standard IPv4 formats
ipv4_integer = r'\b\d{8,10}\b'                      # 32-bit decimal integer
ipv4_hex = r'\b0x[0-9A-Fa-f]{8}\b'                  # Hexadecimal
ipv4_octal = r'\b0[0-7]+\.[0-7]+\.[0-7]+\.[0-7]+\b' # Octal

# Combine all regexes
all_patterns = f"{ipv4_pattern}|{cidr_pattern}|{ipv6_full}|{ipv6_compressed}|{ipv6_mapped_ipv4}|{ipv4_integer}|{ipv4_hex}|{ipv4_octal}"
matches = re.findall(all_patterns, text_data)

# Normalize and validate
valid_ips = []

for match in matches:
    ip = next(filter(None, match))  # Get non-empty group from match tuple
    try:
        # Handle hex (e.g., 0xC0A80101)
        if ip.startswith("0x"):
            ip = ipaddress.IPv4Address(int(ip, 16)).compressed
        # Handle decimal integer
        elif re.fullmatch(ipv4_integer, ip):
            ip = ipaddress.IPv4Address(int(ip)).compressed
        # Handle octal
        elif re.fullmatch(ipv4_octal, ip):
            parts = [str(int(part, 8)) for part in ip.split('.')]
            ip = '.'.join(parts)
            ipaddress.IPv4Address(ip)  # validate
        else:
            ipaddress.ip_network(ip, strict=False)
        valid_ips.append(ip)
    except ValueError:
        continue


print(f"Total valid IP addresses (all formats): {len(valid_ips)}")

Total valid IP addresses (all formats): 0


#### RESULT- There are no IPV6 or IPV4 or any of the above checked IP address found in the file

## Q.No A.8

#### What is the total number of hashtags across all these tweets?

In [133]:
hashtags = re.findall(r'#\w+', text_data)
print(f"Total count of hashtags across all rows/records: {len(hashtags)}")

Total count of hashtags across all rows/records: 830


#### RESULT- 830 Hashtags were found across the tweets in the dataset

# Notebook: PartA_Q1andQ4.ipynb

In [16]:
#Importing Libraries
import pandas as pd
import numpy as np
import re

In [3]:
#Import Dataset
filelocation = 'Data.csv'
data = pd.read_csv(filelocation)

In [4]:
print(data.shape)

(7339, 1)


In [8]:
print(data.head(15))

                                                 Data
0   Watch or listen live weekdays at 8:30am MT at ...
1   Watch or listen live weekdays at 8:30am MT at ...
2                Chubby And Hot, Always Stir The Pot!
3                Chubby And Hot, Always Stir The Pot!
4   Journalist, publisher of Rebel News — telling ...
5   Edmonton City Councillor for #WardMétis. #YEG ...
6   Albertan. Mom to children & dogs. Wife. Friend...
7   Albertan. Mom to children & dogs. Wife. Friend...
8   Albertan. Mom to children & dogs. Wife. Friend...
9   Albertan. Mom to children & dogs. Wife. Friend...
10  Former progressive conservative specializing i...
11  Former progressive conservative specializing i...
12  I'm Kuno, a Gov't qualified service dog in Alb...
13  I'm Kuno, a Gov't qualified service dog in Alb...
14  Albertan. Mom to children & dogs. Wife. Friend...


In [7]:
##checking for NA or NULL values
is_null = data.isnull().sum()
print(is_null)
is_na = data.isna().sum()
print(is_na)

Data    0
dtype: int64
Data    0
dtype: int64


In [10]:
##checking for duplicates
duplicated = data.duplicated()
print(duplicated.sum())

5353


In [11]:
##Remove duplicates
data_v1 = data.drop_duplicates()
print(data_v1.shape)

(1986, 1)


In [15]:
data_v1.to_csv('C:\\Users\\admin\\Downloads\\Cleaned_Data.csv')

In [22]:
print(data_v1.head(5))

                                                Data
0  Watch or listen live weekdays at 8:30am MT at ...
2               Chubby And Hot, Always Stir The Pot!
4  Journalist, publisher of Rebel News — telling ...
5  Edmonton City Councillor for #WardMétis. #YEG ...
6  Albertan. Mom to children & dogs. Wife. Friend...


Q.No.1 How many records have a date that is expressed without using
alphabets?

In [23]:
## checking for dates in numeric formats
date_pattern = r'\b(?:\d{1,4}[-/.]){2}\d{1,4}\b'
# Check each row for a date using regex (returns True/False)
data_v1['Has_Date'] = data_v1['Data'].str.contains(date_pattern)
data_v1['Extracted_Date'] = data_v1['Data'].str.extract(f'({date_pattern})')

# Show result
print(data_v1)



                                                   Data  Has_Date  \
0     Watch or listen live weekdays at 8:30am MT at ...     False   
2                  Chubby And Hot, Always Stir The Pot!     False   
4     Journalist, publisher of Rebel News — telling ...     False   
5     Edmonton City Councillor for #WardMétis. #YEG ...     False   
6     Albertan. Mom to children & dogs. Wife. Friend...     False   
...                                                 ...       ...   
7329          Nerdy guy who likes hockey and punk rock!     False   
7330  Song Writer-Lover of the Creator, Lorna, Bear,...     False   
7332  Radical Downtown Urbanite\nThe other place: @c...     False   
7333  Formerly Agent for The Attorney General of Can...     False   
7334  Snow Valley Racing delivers competitive ski ra...     False   

     Extracted_Date  
0               NaN  
2               NaN  
4               NaN  
5               NaN  
6               NaN  
...             ...  
7329            N

C:\Users\Admin\AppData\Local\Temp\ipykernel_8384\2730574019.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_v1['Has_Date'] = data_v1['Data'].str.contains(date_pattern)
C:\Users\Admin\AppData\Local\Temp\ipykernel_8384\2730574019.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_v1['Extracted_Date'] = data_v1['Data'].str.extract(f'({date_pattern})')


In [25]:

##checking whether the row has dates 
# Filter rows where a date was detected (Has_Date == True)
true_cases = data_v1[data_v1['Has_Date'] == True]
# Display number of true cases and sample results
print("Total rows with numeric dates:", len(true_cases))
print(true_cases.head(16))

Total rows with numeric dates: 16
                                                   Data  Has_Date  \
196   Alberta's NDP & @RachelNotley: A Better Future...      True   
201   Official Alberta Parks Twitter account. Monito...      True   
1185  Small town RV Dealership in Acme, Alberta with...      True   
1410  Serving #strathco, including #shpk and the sur...      True   
1959  For emergencies, please call Report A Poacher ...      True   
2332  Listen live on WILD 95-5 FM. App: Lethbridge N...      True   
2507  Represent. Promote. Educate. Aggregate is the ...      True   
3136  For over 15 years, RV Direct Insurance has bee...      True   
3393  Account is not monitored 24/7. Questions: call...      True   
3609  Saskatchewan's 1st 24 hour commercial Christia...      True   
3637  Nestled in the heart of the Rockies! Stop in, ...      True   
3682  Providing RV repair, parts & accessories, 24/7...      True   
4677  Leaders in highway & bridge maintenance. Maint...      True   


These are phone or toll free numbers not real dates.

In [26]:
## removing phone or toll free numbers
phone_pattern = r'^\d{3}[-.]\d{3}[-.]\d{4}$'
toll_free_pattern = r'^1[-.]\d{3}[-.]\d{3}$'
emergency_pattern = r'^\d{1,2}[-.]\d{1,2}[-.]\d{1,2}$'  

# Remove matches that are likely false positives
df_cleaned = data_v1[
    ~data_v1['Extracted_Date'].str.match(phone_pattern, na=False) &
    ~data_v1['Extracted_Date'].str.match(toll_free_pattern, na=False) &
    ~data_v1['Extracted_Date'].str.match(emergency_pattern, na=False)
]

# Step 5: Filter for valid dates only
valid_dates = df_cleaned[df_cleaned['Has_Date'] == True]

# View result
print("Number of dates found:", len(valid_dates))
print(valid_dates[['Data', 'Extracted_Date']].head())

Number of dates found: 0
Empty DataFrame
Columns: [Data, Extracted_Date]
Index: []


Q.No.4 How many tweets contain one of these emojis :), :D, ;), :P?

In [28]:
# Define regex pattern for the emojis
emoji_pattern = r'(:\)|:D|;\)|:P)'
# Check if each row contains any of the emojis
data_v1['Has_Emoji'] = data_v1['Data'].astype(str).str.contains(emoji_pattern)
print(data_v1[data_v1['Has_Emoji']].head())
# Count the number of matching tweets
emoji_count = data_v1['Has_Emoji'].sum()
print("Number of tweets containing emojis :), :D, ;), or :P =", emoji_count)



                                                   Data  Has_Date  \
2594  Proud parent, photographer, academic book edit...     False   
3259  Animal Advocate, Dog Lover, Rhino Adorer, Vege...     False   
5082  Helping Pet-Parents in Lethbridge & Area reuni...     False   
5763  We're committed to bringing a safe & secure me...     False   

     Extracted_Date  Has_Emoji  
2594            NaN       True  
3259            NaN       True  
5082            NaN       True  
5763            NaN       True  
Number of tweets containing emojis :), :D, ;), or :P = 4


C:\Users\Admin\AppData\Local\Temp\ipykernel_8384\1090961685.py:4: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  data_v1['Has_Emoji'] = data_v1['Data'].astype(str).str.contains(emoji_pattern)
C:\Users\Admin\AppData\Local\Temp\ipykernel_8384\1090961685.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_v1['Has_Emoji'] = data_v1['Data'].astype(str).str.contains(emoji_pattern)


In [29]:
##Cross checking the results using re module
# Define the regex pattern for the emojis
emoji_pattern = re.compile(r'(:\)|:D|;\)|:P)')

# Define a function to check for emoji presence in a string
def contains_emoji(text):
    return bool(emoji_pattern.search(str(text)))  # convert to string and search

# Apply the function to each row in the Data column
data_v1['Has_Emoji_1'] = data_v1['Data'].apply(contains_emoji)

# Count and show results
emoji_count = data_v1['Has_Emoji_1'].sum()
print("Number of tweets with emojis :), :D, ;), or :P =", emoji_count)




Number of tweets with emojis :), :D, ;), or :P = 4


C:\Users\Admin\AppData\Local\Temp\ipykernel_8384\1803070784.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_v1['Has_Emoji_1'] = data_v1['Data'].apply(contains_emoji)


In [34]:
emoji_tweets = data_v1[data_v1['Has_Emoji']]['Data']
pd.set_option('display.max_colwidth', None)
print(emoji_tweets)

2594    Proud parent, photographer, academic book editor, and tryer of new recipes. I'm not a cannibal or an apocalypse survivor but I'll be playing one on TV ;)
3259               Animal Advocate, Dog Lover, Rhino Adorer, Vegetarian, Dancer and Yogini:) Monarchist, Loved HMTQ, Tea Time but definitely no Sugar in My Tea;)
5082                                                                                    Helping Pet-Parents in Lethbridge & Area reunite with their Lost Pets. :)
5763        We're committed to bringing a safe & secure method of compressed gas transportation to the market. Oh ya and we also love to just meet cool people :)
Name: Data, dtype: object


The Number of tweets with emojis = 4

# Notebook: PartA_Q2_Q3.ipynb

### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import re

### Dataset Overview

In [2]:
# Load the Data.CSV file
file_path = "Data.csv"
data = pd.read_csv(file_path, keep_default_na=True)
print(data.info)
print("Dataset loaded Successfully")

<bound method DataFrame.info of                                                    Data
0     Watch or listen live weekdays at 8:30am MT at ...
1     Watch or listen live weekdays at 8:30am MT at ...
2                  Chubby And Hot, Always Stir The Pot!
3                  Chubby And Hot, Always Stir The Pot!
4     Journalist, publisher of Rebel News — telling ...
...                                                 ...
7334  Snow Valley Racing delivers competitive ski ra...
7335  New twin mom\n\nMorning News Anchor @citynewsc...
7336  Residents advocating for the💚of the Park.\n\nP...
7337  Driven by family, diversity, equity, inclusion...
7338      Photographer, Oilers Fan #photography  #yeg .

[7339 rows x 1 columns]>
Dataset loaded Successfully


In [5]:
data.head(10)

,Data
0,Watch or listen live weekdays at 8:30am MT at ...
1,Watch or listen live weekdays at 8:30am MT at ...
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
4,"Journalist, publisher of Rebel News — telling ..."
5,Edmonton City Councillor for #WardMétis. #YEG ...
6,Albertan. Mom to children & dogs. Wife. Friend...
7,Albertan. Mom to children & dogs. Wife. Friend...
8,Albertan. Mom to children & dogs. Wife. Friend...
9,Albertan. Mom to children & dogs. Wife. Friend...


#### Data Pre-processing

In [6]:
# Remove duplicate rows
df = data.drop_duplicates()

In [7]:
# Check for null (missing) values
null_counts = df.isnull().sum()
print(null_counts)

Data    0
dtype: int64


In [50]:
# Convert the "Data" column to string
df_string= df["Data"].astype(str)
df_string.info()

<class 'pandas.core.series.Series'>
Index: 1986 entries, 0 to 7334
Series name: Data
Non-Null Count  Dtype 
--------------  ----- 
1986 non-null   object
dtypes: object(1)
memory usage: 31.0+ KB


### Part A 
#### Q2. How many records have a word starting with the letter “w”?

In [108]:
# Defining a regular expression to include words starting with "w" (case sensitive)
match_pattern_w = r'\bw[a-zA-Z]*\b'

In [53]:
# Filter matching rows
matching_rows = df_string[df_string.str.contains(match_pattern_w, regex=True)]

# A sample output of 5 matched rows 
sample_matches = matching_rows.head(5).tolist()

for i, row in enumerate(sample_matches, 1):
    print(f"{i}: {row}")

1: Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
2: I'm Kuno, a Gov't qualified service dog in Alberta, Canada. My job is taking care of my person, a bilateral amputee, #weatherdog & Spokesdog for the disabled
3: Default avatar or locked followers will be removed (locked accounts can DM to discuss). Crypto-bros will be blocked. DianeMariePosts@mstdn.ca|Same on Post.
4: Dealing with Bipolar disorder one day at a time.Drug overdose survivor.#thygoddostexist Feeling good is good enough!🙂All photographs were taken by me.🚫DM‼️
5: Proud dad. Happy husband. Union member. Works in comms. I write about Alberta politics and history at Daveberta.Substack.com | ⛷🚵 he/him


In [109]:
# Count of records matching the defined pattern
count_w = df_string.str.contains(match_pattern_w, regex=True).sum()
print("Records with a word starting with 'w':", count_w)

Records with a word starting with 'w': 637


#### Result & Observation(s)

### 637 Records

It is observed that the hashtags are also taken into consideration as '#' is not a word character and is ignored by the regex engine (For example : #weatherdog in line 2) 

#### Q3. How many records make a word that starts with an alphabet and is not a URL?

In [115]:
# Defining a pattern to find all words starting with an alphabet
word_pattern = re.compile(r'\b[a-zA-Z][\w-]*\b')

# List of common domain extensions to filter
domain_extensions = ['com', 'org', 'net', 'io', 'gov', 'edu', 'ca', 'co', 'us', 'uk', 'info', 'biz', 'me']

# Pattern to detect domains
domain_regex = re.compile(r'\b[\w.-]+\.(?:' + '|'.join(domain_extensions) + r')\b', re.IGNORECASE)

In [117]:
# Function to check if record contains a valid word that is not a domain or URL
def has_valid_non_url_word(text):
    if not isinstance(text, str):
        return False
    if domain_regex.search(text):
        return False
    words = word_pattern.findall(text)
    return len(words) > 0

# Filter records
valid_records = df_string[df_string.apply(has_valid_non_url_word)]
valid_records.head()

2                  Chubby And Hot, Always Stir The Pot!
4     Journalist, publisher of Rebel News — telling ...
5     Edmonton City Councillor for #WardMétis. #YEG ...
6     Albertan. Mom to children & dogs. Wife. Friend...
10    Former progressive conservative specializing i...
Name: Data, dtype: object

In [118]:
# Print number of valid records
print("Count of records that make a word that starts with an alphabet and is not a URL : ", len(valid_records))

Count of records that make a word that starts with an alphabet and is not a URL :  1787


#### Result & Observation(s)

### 1787 Records
The regular expression strictly filters all urls from the given data, including most common domains. 